File: legacy/apple-prototype/Sources/CRMCore/Import.swift
import Foundation

public struct CSVTable: Equatable, Sendable {
    public let headers: [String]
    public let rows: [[String]]
}
public enum CSVParser {
    public static func parse(_ data: Data) throws -> CSVTable {
        guard data.count <= 10 * 1024 * 1024 else { throw CRMError.validation("Import files must be 10 MB or smaller.") }
        let text: String
        if data.starts(with: [0xff, 0xfe]) || data.starts(with: [0xfe, 0xff]) {
            guard let value = String(data: data, encoding: .utf16) else { throw CRMError.unsupportedEncoding }; text = value
        } else {
            guard let value = String(data: data, encoding: .utf8) else { throw CRMError.unsupportedEncoding }; text = value
        }
        let chars = Array(text.trimmingCharacters(in: CharacterSet(charactersIn: "\u{feff}")))
        var rows: [[String]] = []; var row: [String] = []; var field = ""
        var quoted = false; var quoteClosed = false; var i = 0
        while i < chars.count {
            let c = chars[i]
            if quoted {
                if c == "\"" {
                    if i + 1 < chars.count && chars[i + 1] == "\"" { field.append("\""); i += 1 }
                    else { quoted = false; quoteClosed = true }
                } else { field.append(c) }
            } else if c == "," { row.append(field); field = ""; quoteClosed = false }
            else if c == "\n" || c == "\r" || c == "\r\n" {
                row.append(field); rows.append(row); row = []; field = ""; quoteClosed = false
                if c == "\r", i + 1 < chars.count, chars[i + 1] == "\n" { i += 1 }
            } else if c == "\"" {
                guard field.isEmpty && !quoteClosed else { throw CRMError.malformedCSV }; quoted = true
            } else {
                guard !quoteClosed else { throw CRMError.malformedCSV }; field.append(c)
            }
            i += 1
        }
        guard !quoted else { throw CRMError.malformedCSV }
        if !row.isEmpty || !field.isEmpty || quoteClosed { row.append(field); rows.append(row) }
        guard let first = rows.first else { throw CRMError.malformedCSV }
        let headers = first.map { $0.trimmingCharacters(in: .whitespacesAndNewlines) }
        guard !headers.isEmpty, headers.allSatisfy({ !$0.isEmpty }), Set(headers.map { $0.lowercased() }).count == headers.count else { throw CRMError.malformedCSV }
        let body = Array(rows.dropFirst()).filter { !$0.allSatisfy(\.isEmpty) }
        guard body.allSatisfy({ $0.count == headers.count }) else { throw CRMError.malformedCSV }
        return CSVTable(headers: headers, rows: body)
    }
}
public struct ColumnMapping: Sendable {
    public var name: String
    public var email: String?
    public var phone: String?
    public init(name: String, email: String? = nil, phone: String? = nil) { self.name = name; self.email = email; self.phone = phone }
}
public struct DuplicateMatch: Equatable, Sendable {
    public let personID: UUID
    public let confidence: Double
    public let reasons: [String]
}
public struct ImportRow: Identifiable, Sendable {
    public let id: Int
    public let person: Person
    public let errors: [String]
    public let duplicates: [DuplicateMatch]
}
public enum ImportChoice: Equatable, Sendable { case skip, create, merge(UUID) }
public struct ImportPreview: Sendable {
    public let id: UUID
    public let rows: [ImportRow]
    public let at: Date
    public var validCount: Int { rows.filter { $0.errors.isEmpty }.count }
}
public enum ContactNormalization {
    public static func email(_ value: String) -> String { value.trimmingCharacters(in: .whitespacesAndNewlines).lowercased() }
    public static func phone(_ value: String) -> String {
        let trimmed = value.trimmingCharacters(in: .whitespacesAndNewlines)
        let permitted = CharacterSet(charactersIn: "+0123456789 ()-.")
        guard trimmed.unicodeScalars.allSatisfy({ permitted.contains($0) }) else { return trimmed }
        let digits = trimmed.filter(\.isNumber)
        return (trimmed.hasPrefix("+") ? "+" : "") + digits
    }
}
public struct ImportEngine: Sendable {
    public init() {}
    public func preview(_ data: Data, mapping: ColumnMapping, in db: Database, at: Date) throws -> ImportPreview {
        let table = try CSVParser.parse(data)
        guard let nameIndex = table.headers.firstIndex(of: mapping.name) else { throw CRMError.validation("Map the name column.") }
        for column in [mapping.email, mapping.phone].compactMap({ $0 }) where !table.headers.contains(column) { throw CRMError.validation("A mapped column does not exist.") }
        let emailIndex = mapping.email.flatMap { table.headers.firstIndex(of: $0) }; let phoneIndex = mapping.phone.flatMap { table.headers.firstIndex(of: $0) }
        var candidates = db.visiblePeople
        var rows: [ImportRow] = []
        for (index, values) in table.rows.enumerated() {
            var person = Person(name: values[nameIndex].trimmingCharacters(in: .whitespacesAndNewlines), at: at, source: .imported)
            let email = emailIndex.map { ContactNormalization.email(values[$0]) } ?? ""
            let phone = phoneIndex.map { ContactNormalization.phone(values[$0]) } ?? ""
            person.emails = email.isEmpty ? [] : [email]; person.phones = phone.isEmpty ? [] : [phone]
            var errors: [String] = []
            if person.name.isEmpty { errors.append("Name is required") }
            if !email.isEmpty && (email.split(separator: "@").count != 2 || email.contains(" ")) { errors.append("Email format needs review") }
            if !phone.isEmpty && phone.filter(\.isNumber).count < 7 { errors.append("Phone number needs review") }
            let matches = candidates.compactMap { candidate -> DuplicateMatch? in
                var reasons: [String] = []
                if !email.isEmpty && candidate.emails.map(ContactNormalization.email).contains(email) { reasons.append("Exact normalized email") }
                if !phone.isEmpty && candidate.phones.map(ContactNormalization.phone).contains(phone) { reasons.append("Exact normalized phone") }
                if candidate.name.compare(person.name, options: [.caseInsensitive, .diacriticInsensitive]) == .orderedSame { reasons.append("Matching name") }
                guard !reasons.isEmpty else { return nil }
                return DuplicateMatch(personID: candidate.id, confidence: reasons.contains(where: { $0.hasPrefix("Exact") }) ? 0.98 : 0.65, reasons: reasons)
            }.sorted { $0.confidence == $1.confidence ? $0.personID.uuidString < $1.personID.uuidString : $0.confidence > $1.confidence }
            rows.append(ImportRow(id: index + 2, person: person, errors: errors, duplicates: matches))
            if errors.isEmpty { candidates.append(person) }
        }
        return ImportPreview(id: UUID(), rows: rows, at: at)
    }
}
public extension CRMService {
    func commitImport(_ preview: ImportPreview, choices: [Int: ImportChoice], at: Date, operationID: UUID) throws {
        try transact(target: preview.id, action: "import.committed", at: at, operationID: operationID) { db in
            guard db.imports[preview.id] == nil else { throw CRMError.duplicate }
            var created: [UUID: Person] = [:]
            for row in preview.rows {
                guard let choice = choices[row.id] else { throw CRMError.validation("Choose create, skip, or merge for each row.") }
                if choice == .skip { continue }
                guard row.errors.isEmpty else { throw CRMError.validation("Fix or skip invalid import rows.") }
                switch choice {
                case .skip: break
                case .create:
                    guard db.people[row.person.id] == nil else { throw CRMError.duplicate }
                    db.people[row.person.id] = row.person; created[row.person.id] = row.person
                case .merge(let targetID):
                    // Explicit merge only fills contact methods; no controlled field is replaced.
                    var target = try Self.person(targetID, in: db)
                    target.emails = Array(Set(target.emails + row.person.emails)).sorted()
                    target.phones = Array(Set(target.phones + row.person.phones)).sorted()
                    db.people[targetID] = target
                    Self.append(&db, id: targetID, at: at, kind: .note, summary: "Imported contact methods merged", operation: operationID)
                }
            }
            // Imported values acquire revision 0 on commit. Rollback validates exact material state.
            for id in created.keys { created[id]?.metadata.updatedAt = at }
            db.imports[preview.id] = ImportBatch(id: preview.id, at: at, created: created)
        }
    }
    func rollbackImport(_ batchID: UUID, at: Date, operationID: UUID) throws {
        try transact(target: batchID, action: "import.rolledBack", at: at, operationID: operationID) { db in
            guard var batch = db.imports[batchID], batch.rolledBackAt == nil else { throw CRMError.notFound }
            for (id, original) in batch.created {
                guard db.people[id] == original,
                      !db.activities.values.contains(where: { $0.personID == id }),
                      !db.tasks.values.contains(where: { $0.personID == id }),
                      !db.referrals.values.contains(where: { $0.personID == id || $0.referrerID == id }),
                      !db.files.values.contains(where: { $0.personID == id }),
                      !db.edges.values.contains(where: { $0.fromID == id || $0.toID == id }) else { throw CRMError.rollbackConflict }
            }
            for id in batch.created.keys { db.people.removeValue(forKey: id) }
            batch.rolledBackAt = at; db.imports[batchID] = batch
        }
    }
}

File: web/src/lib/server/import.ts
import type { Snapshot, ImportPreview, ImportRow } from "@/contracts/crm";
import { ApiError } from "./errors";
export function parseCSV(csv: string): { headers: string[]; rows: string[][] } {
  if (Buffer.byteLength(csv) > 1024 * 1024)
    throw new ApiError("IMPORT_INVALID");
  const text = csv.replace(/^\uFEFF/, "");
  let row: string[] = [],
    field = "",
    quoted = false,
    closed = false;
  const rows: string[][] = [];
  for (let i = 0; i < text.length; i++) {
    const c = text[i];
    if (quoted) {
      if (c === '"') {
        if (text[i + 1] === '"') {
          field += '"';
          i++;
        } else {
          quoted = false;
          closed = true;
        }
      } else field += c;
    } else if (c === ",") {
      row.push(field);
      field = "";
      closed = false;
    } else if (c === "\r" || c === "\n") {
      row.push(field);
      rows.push(row);
      row = [];
      field = "";
      closed = false;
      if (c === "\r" && text[i + 1] === "\n") i++;
    } else if (c === '"') {
      if (field || closed) throw new ApiError("IMPORT_INVALID");
      quoted = true;
    } else {
      if (closed) throw new ApiError("IMPORT_INVALID");
      field += c;
    }
  }
  if (quoted) throw new ApiError("IMPORT_INVALID");
  if (row.length || field || closed) {
    row.push(field);
    rows.push(row);
  }
  const headers = rows.shift()?.map((x) => x.trim());
  if (
    !headers?.length ||
    headers.some((x) => !x) ||
    new Set(headers.map((x) => x.toLowerCase())).size !== headers.length
  )
    throw new ApiError("IMPORT_INVALID");
  const body = rows.filter((r) => r.some(Boolean));
  if (body.length > 1000 || body.some((r) => r.length !== headers.length))
    throw new ApiError("IMPORT_INVALID");
  return { headers, rows: body };
}
export const normalizeEmail = (value: string) => value.trim().toLowerCase();
export function normalizePhone(value: string) {
  const trimmed = value.trim();
  return /^[+\d ().-]+$/.test(trimmed)
    ? (trimmed.startsWith("+") ? "+" : "") + trimmed.replace(/\D/g, "")
    : trimmed;
}
export function previewImport(
  csv: string,
  mapping: { name: string; email?: string; phone?: string },
  db: Snapshot,
): ImportPreview {
  const table = parseCSV(csv);
  const ni = table.headers.indexOf(mapping.name),
    ei = mapping.email ? table.headers.indexOf(mapping.email) : -1,
    pi = mapping.phone ? table.headers.indexOf(mapping.phone) : -1;
  if (
    new Set(Object.values(mapping)).size !== Object.values(mapping).length ||
    ni < 0 ||
    (mapping.email && ei < 0) ||
    (mapping.phone && pi < 0)
  )
    throw new ApiError("IMPORT_INVALID");
  const names = new Map<string, Set<string>>(),
    emails = new Map<string, Set<string>>(),
    phones = new Map<string, Set<string>>();
  function index(map: Map<string, Set<string>>, key: string, id: string) {
    if (!key) return;
    const values = map.get(key) ?? new Set();
    values.add(id);
    map.set(key, values);
  }
  for (const p of db.people.filter((p) => !p.archivedAt && !p.mergedInto)) {
    index(names, p.name.toLowerCase(), p.id);
    p.emails.forEach((e) => index(emails, normalizeEmail(e), p.id));
    p.phones.forEach((pn) =>
      index(phones, normalizePhone(pn).replace(/\D/g, ""), p.id),
    );
  }
  const rows: ImportRow[] = table.rows.map((values, i) => {
    const name = values[ni].trim(),
      email = ei >= 0 ? normalizeEmail(values[ei]) : "",
      phone = pi >= 0 ? normalizePhone(values[pi]) : "";
    const errors: string[] = [];
    if (!name || name.length > 200)
      errors.push("Name is required and must be 200 characters or fewer");
    if (
      email &&
      (email.length > 320 || !/^[^\s@]+@[^\s@]+\.[^\s@]+$/.test(email))
    )
      errors.push("Email format needs review");
    if (
      phone &&
      (!/^[+\d ().-]+$/.test(phone) ||
        phone.replace(/\D/g, "").length < 7 ||
        phone.length > 40)
    )
      errors.push("Phone format needs review");
    const matches = new Map<string, string[]>();
    for (const [map, key, reason] of [
      [names, name.toLowerCase(), "Matching name"],
      [emails, email, "Exact normalized email"],
      [phones, phone.replace(/\D/g, ""), "Exact normalized phone"],
    ] as const) {
      for (const id of map.get(key) ?? []) {
        const reasons = matches.get(id) ?? [];
        reasons.push(reason);
        matches.set(id, reasons);
      }
    }
    const duplicates = [...matches]
      .map(([id, reasons]) => ({
        id,
        reasons,
        confidence: reasons.some((r) => r.startsWith("Exact")) ? 0.98 : 0.65,
      }))
      .sort((a, b) => b.confidence - a.confidence || a.id.localeCompare(b.id));
    if (!errors.length) {
      const id = `row:${i + 2}`;
      index(names, name.toLowerCase(), id);
      index(emails, email, id);
      index(phones, phone.replace(/\D/g, ""), id);
    }
    return { row: i + 2, name, email, phone, errors, duplicates };
  });
  return {
    rows,
    validCount: rows.filter((x) => !x.errors.length).length,
    duplicateCount: rows.filter((x) => x.duplicates.length).length,
  };
}
